### PR TITLE
[Feature #68029]: Quản lý sản phẩm

### PR DESCRIPTION
**Pull request by BuiMinhDung711**
## Related Tickets:
- [#68036](https://edu-redmine.sun-asterisk.vn/issues/68036)
- [#68038](https://edu-redmine.sun-asterisk.vn/issues/68038)
- [#68042](https://edu-redmine.sun-asterisk.vn/issues/68042)

## WHAT this PR does?
- Thêm, sửa, xóa (mềm) một sản phẩm

## Notes Impacted Areas:
- ```app/Http/Controllers/ProductController.php```
- ```app/Http/Livewire/ProductsGridView.php```
- ```app/Http/Requests/Product```
- ```app/View```
- ```resources/views/product```
- ```resources/views/vendor/laravel-views/grid-view/grid-view-item.blade.php```
- ```routes/web.php```

## Attachments:
- image
- image
-

### CODE REVIEW

**Comment by BuiMinhDung711** [comment]
ready

File: app/Http/Controllers/ProductController.php [hunk 1]
```diff
@@ -4,14 +4,27 @@
...
 {
+    protected function storeImage(Request $request)
+    {
+        $fileName = time() . uniqid() . '.' . $request->get('brand')
+                    . $request->get('name') . '.'
+                    . $request->file('big_image')->extension();
+        $path = $request->file('big_image')->storeAs('public/images/Product', $fileName);
+        return 'storage/images/Product/' . $fileName;
+    }
+
```
Comment (ghost) [inline app/Http/Controllers/ProductController.php]: use Storage:: facade

File: app/Http/Controllers/ProductController.php [hunk 2]
```diff
@@ -56,8 +69,53 @@ public function create()
...
+            'brand' => $validatedData['brand'],
+            'name' => $validatedData['name'],
+        ]);
+
+        $productId = Product::where('product_uuid', $validatedData['product_uuid'])->first()->id;
+
+        if ($request->hasFile('big_image')) {
+            $fileName = time() . uniqid() . '.' . $request->get('brand')
+                        . $request->get('name') . '.'
+                        . $request->file('big_image')->extension();
```
Comment (ghost) [inline app/Http/Controllers/ProductController.php]: indent 4 spaces

File: app/Http/Controllers/ProductController.php [hunk 3]
```diff
@@ -56,8 +69,53 @@ public function create()
...
+            $path = $request->file('big_image')->storeAs('public/images/Product', $fileName);
+            $imageUrl = 'storage/images/Product/' . $fileName;
+            $validatedData['big_image'] = $imageUrl;
+            ProductMedia::create([
+                'product_id' => $productId,
+                'type' => 'big image',
+                'media_link' => $validatedData['big_image'],
+            ]);
+        }
```
Comment (ghost) [inline app/Http/Controllers/ProductController.php]: use Storage:: facade

File: app/Http/Controllers/ProductController.php [hunk 4]
```diff
@@ -102,20 +160,66 @@ public function edit(Product $product)
...
+            'brand' => $validatedData['brand'],
+            'name' => $validatedData['name'],
+        ]);
+
+        if ($request->hasFile('big_image')) {
+            ProductMedia::where('product_id', $product->id)->where('type', 'big image')->delete();
+
+            $fileName = time() . uniqid() . '.' . $request->get('brand')
+                        . $request->get('name') . '.'
+                        . $request->file('big_image')->extension();
```
Comment (ghost) [inline app/Http/Controllers/ProductController.php]: indent 4 spaces

File: app/Http/Livewire/ProductsGridView.php [hunk 5]
```diff
@@ -42,19 +49,37 @@ public function repository(): Builder
...
+        $big_image = 'storage/images/logo/no-product-image.png';
+        $product_media = $product->productMedias->where('type', 'big image')->first();
+        $description = '';
+        if ($product->deleted_at) {
+            $description = "This product has been deleted. It's unavailable now.";
+        } else {
+            if ($product->productInStocks->count() === 0) {
+                $description = "This product is unavailable now. You can add product's quantity.";
+            } else {
+                $description = "This product is out of stock. You can update product's quantity.";
```
Comment (ghost) [inline app/Http/Livewire/ProductsGridView.php]: localization

File: app/Http/Livewire/ProductsGridView.php [hunk 6]
```diff
@@ -42,19 +49,37 @@ public function repository(): Builder
      */
     public function card($product)
     {
-        $big_image = $product->productMedias->where('type', 'big image')->first()->media_link;
-        
-        if ($big_image === null) {
-            $big_image = '';
+        $big_image = 'storage/images/logo/no-product-image.png';
```
Comment (ghost) [inline app/Http/Livewire/ProductsGridView.php]: put in config, ảnh default hệ thống để trong storage ntn có đc đẩy lên ko đó?

File: app/Http/Livewire/ProductsGridView.php [hunk 7]
```diff
@@ -42,19 +49,37 @@ public function repository(): Builder
...
+        if ($product->deleted_at) {
+            $description = "This product has been deleted. It's unavailable now.";
+        } else {
+            if ($product->productInStocks->count() === 0) {
+                $description = "This product is unavailable now. You can add product's quantity.";
+            } else {
+                $description = "This product is out of stock. You can update product's quantity.";
+                foreach ($product->productInStocks as $productInStock) {
+                    if ($productInStock->quantity != 0) {
+                        $description = "This product is available.";
```
Comment (ghost) [inline app/Http/Livewire/ProductsGridView.php]: localization

File: app/Http/Requests/Product/StoreProductRequest.php [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+     *
+     * @return array
+     */
+    public function rules()
+    {
+        return [
+            'brand' => ['required', 'string', 'max:255'],
+            'name' => ['required', 'string', 'max:255'],
+            'big_image' => ['image'],
+            'small_image.*' =>['image'],
```
Comment (ghost) [inline app/Http/Requests/Product/StoreProductRequest.php]:
```suggestion
            'small_image.*' => ['image'],
```

File: app/Http/Requests/Product/UpdateProductRequest.php [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+     *
+     * @return array
+     */
+    public function rules()
+    {
+        return [
+            'brand' => ['required', 'string', 'max:255'],
+            'name' => ['required', 'string', 'max:255'],
+            'big_image' => ['image'],
+            'small_image.*' =>['image'],
```
Comment (ghost) [inline app/Http/Requests/Product/UpdateProductRequest.php]:
```suggestion
            'small_image.*' => ['image'],
```

File: app/View/Actions/UpdateProductAction.php [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+use LaravelViews\Actions\Action;
+use LaravelViews\Views\View;
+
+class UpdateProductAction extends Action
+{
+    /**
+     * Any title you want to be displayed
+     * @var String
+     * */
+    public $title = "Update this product";
```
Comment (ghost) [inline app/View/Actions/UpdateProductAction.php]: localization

File: routes/web.php [hunk 11]
```diff
@@ -58,6 +73,6 @@
         'auth',
         'verified',
         'role:admin',
-]);
+    ]);
```
Comment (ghost) [inline routes/web.php]: ? random indent?

**Comment by BuiMinhDung711** [comment]
ready

**Pull request merged**